enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,226 @@
+
+#### Story 3
+- As an operator of an OCP Dedicated Managed Platform, I want to give users flexibility to add as many workers nodes as they want or to enable autoscaling on worker nodes so I need to have ability to resize the control plane instances in a declarative, automated and seamless manner to react quickly to cluster growth.
+
+#### Story 4
+- As a SRE, I want to have consumable API primitives in place to resize Control Plane compute resources so I can develop upper level automation tooling atop. E.g Control Plane autoresizing to support a severe growth peak of the number of worker nodes.
+
+#### Story 5
+- As an operator, I want my machine API cluster infrastructure to be self healing so I want faulty nodes to be remediated automatically for safety scenarios. This includes having self healing Control Plane machines.
+
+#### Story 6
+- As a multi cluster operator, I want to have a universal user experience for managing the Control Plane in a declarative manner across any cloud provider, bare metal and any flavour of the product that have in common the topology assumed in this doc.
+
+#### Story 7
+- As a developer, I want to be able to deploy the smallest possible cluster to save costs, i.e one instance Control Plane.
+
+### Implementation Details/Notes/Constraints [optional]
+
+To satisfy the goals, motivation and stories above this proposes to let the installer to create MachineSets to own the Machine objects for each failure domain and a Machine Health Checking to monitor the Control Plane machines.
+
+#### Bootstrapping
+Currently during a regular IPI bootstrapping process the installer uses Terraform to create a bootstrapping instance and 3 master instances. Then it creates Machine resources to "adopt" the existing master instances.
+
+Additionally:
+- It will create MachineSets to adopt those machines by looking up known labels (Adopting behaviour already exists in machineSet logic).
+	- `machine.openshift.io/cluster-api-MachineSet": <cluster_name>-<label>-<zone>-controlPlane`
+	- `machine.openshift.io/cluster-api-cluster":    clusterID`
+
+#### Declarative horizontal scaling
+- Each MachineSet can be scaled in/out by changing the replica number or via "scale" subresource.
+- Out of band the Cluster etcd Operator is responsible for watching the existing nodes and taking care of the etcd operational aspects.
+- During scaling in operations any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that might be considered disruptive.
+
+#### Declarative Vertical scaling
+- With MachineSets in place this is semi-automated:
+	- A particular provider property can be changed by any consumer in the MachineSet spec.
+	- This won't affect existing Machines but it would apply for upcoming Machines.
+	- Machine creation with the new spec property can be forced by scaling out or deleting existing machines, e.g `kubectl delete machine name`
```

Comment:
Need to shine a light on this section - stories 1, 3, and 4 (SRE) are not improved much over the existing manual, human-driven process if vertical scaling is not fully automated to the point where we can build automation on top.

Reply:
Trying to provide fully automated vertical rolling upgrade in one single proposal will take quite a lot more friction and time. ([You can notice that here](https://github.com/openshift/enhancements/pull/278)
We don't even support fully rolling upgrade for workers either.

This proposal sets the foundation for any farther automation on top that we may want to follow up with ([See](See https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR51.)).

With this in place nothing would prevent you from start exploring and developing your own automation atop machineSets for e.g the steps in https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR139-R141

I think this is possibly how far we can go as a single deliverable for a release while getting all the teams aligned. We can always discuss a follow up proposal for farther automation including the real use feedback we'll get out of this one.

Reply:
> We don't even support fully rolling upgrade for workers either.

What does this mean?  As far as I know we do.

Reply:
machineSets (so workers) do not know how to do fully automated rolling upgrades.

A change into a machineSet property has not effect over existing machines. For achieving a rolling upgrade you need to put something on top or do some manual steps to recreate machines that will come up with the new properties as described in https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR138-R141.

So this proposal try to align with machineSets and workers.

Reply:
As far as I recall, hive doesn't allow modifying the instance type.  We have had to create new MachinePools to modify instance types for workers and this has had negative impacts on systems making assumption about machineset names as we cannot create a second machineset with the same name.  Support changes to instance type for IPI sounds like a separate enhancement that would then provide the ability for vertical scaling to workers and master.

Reply:
>Support changes to instance type for IPI sounds like a separate enhancement that would then provide the ability for vertical scaling to workers and master.

Yeh, this proposes to align all compute resources on machineSets out of the box. So any farther automation (e.g machineDeployment) can be proposed separately to build atop. With only machineSets the limitations for rolling upgrades are described in https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR138-R141

Reply:
https://github.com/openshift/enhancements/commit/5ef75fdebfd1fb65d2448b21c5dcf5fbb2511fd2